basket/news/tasks.py

```diff
@@ -717,6 +717,250 @@ def record_source_url(email, source_url, newsletter_id):
+    """
+
+    statsd.incr('news.tasks.process_subhub_event.subscription_charge')
+
+    user_data = get_user_data(payee_id=data['customer_id'])
+
+    # the only user identifiable information available is the payment
+    # processor/Stripe ID, so if the user wasn't found by that, there's really
+    # nothing to be done here
+    if user_data:
+        # this is the primary key for the opportunity
+        charge_id = data['charge_id']
+
+        # just in case amount data is missing/malformed
+        try:
+            # amount values are cents - convert to dollars
+            amount_paid = int(data['amount']) / 100
+        except ValueError:
+            amount_paid = 0
+
+        # common data to both events
+        transaction_data = {
+            'PMT_Cust_Id__c': data['customer_id'],
+            'PMT_Subscription_ID__c': data['subscription_id'],
+            'CloseDate': iso_format_unix_timestamp(data['created']),
+            'Billing_Cycle_End__c': iso_format_unix_timestamp(data['period_end']),
+            'Billing_Cycle_Start__c': iso_format_unix_timestamp(data['period_start']),
+            'Amount': amount_paid,
+            'currency__c': data['currency'],
+            'Name': 'Subscription Services',
+            'RecordTypeId': settings.SUBHUB_OPP_RECORD_TYPE,
+            'StageName': 'Closed Won',
+            'Payment_Source__c': 'Stripe',
+            'Invoice_Number__c': data['invoice_id'],
+            'Processors_Fee__c': data['application_fee_amount'],
+        }
+
+        # these keys will be available for the invoice.payment_succeeded event
+        if all(k in data for k in ('brand', 'last4', 'exp_month', 'exp_year')):
+            transaction_data['Credit_Card_Type__c'] = data['brand']
+            transaction_data['Last_4_Digits__c'] = data['last4']
+            transaction_data['Credit_Card_Exp_Month__c'] = data['exp_month']
+            transaction_data['Credit_Card_Exp_Year__c'] = data['exp_year']
+
+        try:
+            # attempt to update the opportunity first (this will happen on the
+            # second event for the current charge)
+            # will raise a SalesforceMalformedRequest if not found
+
+            # if the following succeeds, we don't have to worry about setting
+            # the initial purchase flag - we leave it alone, assuming it was
+            # set properly in the except block below
+            sfdc.opportunity.update('PMT_Transaction_ID__c/{}'.format(charge_id), transaction_data)
+        except sfapi.SalesforceMalformedRequest:
+            # we are creating a new opportunity, so put the charge_id in
+            transaction_data['PMT_Transaction_ID__c'] = charge_id
+
+            # determine if this is an initial or recurring payment
+            try:
+                # look for existing opportunities with the current subscription_id
+                _ = sfdc.opportunity.get_by_custom_id('PMT_Subscription_ID__C', data['subscription_id'])
```

Comment:
Why do you need `_`?

Reply:
Meant to get rid of that - was just coding quickly and wasn't sure if I was going to be using the result. Will remove.